### PR TITLE
*: support memTracker.detach for HashJoin, Apply and IndexLookUp in Close func

### LINKED ISSUE #54005: unexpected memory usage of memory tracker
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
[data.txt](https://github.com/user-attachments/files/15817106/data.txt)
[query.txt](https://github.com/user-attachments/files/15817107/query.txt)
[schema.txt](https://github.com/user-attachments/files/15817108/schema.txt)

``` sql
# mysql -h 10.71.205.177 -uroot -P 4000 --local-infile=1 --comments
source schema.txt
source data.txt
source query.txt
```
After executing this SQL, the heap profile shows that the memory usage related to the tracker is significantly high. This behavior is unexpected.
1. The tracker's memory should not occupy such a large portion.
2. This memory is not being tracked, and therefore it is not included in the total memory usage of a single SQL statement. This can prevent the SQL from being canceled in a timely manner, potentially leading to an Out of Memory risk.

 
截屏2024-06-13 17 38 43

### 4. What is your TiDB version? (Required)
nightly
8d28b23c3c7e40ee475c6d74931e37b728f1b610

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #54005 

Problem Summary:

### What changed and how does it work?
Before:
截屏2024-06-19 10 18 55

After:
截屏2024-06-19 10 06 18

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fixed the issue of abnormally high memory usage caused by memTracker not being detached in HashJoin and IndexLookUp operators when used as the inner side of the Apply operator.
```

### CODE REVIEW

**Review by wshwsh12 — APPROVED** [review]

File: pkg/executor/distsql.go [hunk 1]
```diff
@@ -859,6 +859,7 @@ func (e *IndexLookUpExecutor) Close() error {
 	e.tblWorkerWg.Wait()
 	e.finished = nil
 	e.workerStarted = false
+	e.memTracker.Detach()
```
Comment (guo-shaoge) [inline pkg/executor/distsql.go:862]: move

File: pkg/util/chunk/list.go [hunk 2]
```diff
@@ -45,6 +45,18 @@ type RowPtr struct {
 	RowIdx uint32
 }
 
+// NewListWithMemTracker creates a new List with field types, init chunk size, max chunk size and memory tracker.
+func NewListWithMemTracker(fieldTypes []*types.FieldType, initChunkSize, maxChunkSize int, tracker *memory.Tracker) *List {
+	l := &List{
```
Comment (guo-shaoge) [inline pkg/util/chunk/list.go:50]: maybe call `NewList()` then put `trackers` into it.

**Comment by guo-shaoge** [comment]
other lgtm

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**